packages/errors/src/json-rpc-error.ts

```diff
@@ -0,0 +1,85 @@
+import { SolanaErrorContext } from './context';
+import { SolanaError } from './error';
+import { getSolanaErrorFromTransactionError } from './transaction-error';
+
+interface JsonRpcError {
+    code: number;
+    data?: unknown;
+    message: string;
+}
+
+type TransactionError = string | { [key: string]: unknown };
+
+// Keep in sync with https://github.com/anza-xyz/agave/blob/master/rpc-client-api/src/response.rs
+export interface RpcSimulateTransactionResult {
+    accounts:
+        | ({
+              data:
+                  | string // LegacyBinary
+                  | {
+                        // Json
+                        parsed: unknown;
+                        program: string;
+                        space: number;
+                    }
+                  // Binary
+                  | [encodedBytes: string, encoding: 'base58' | 'base64' | 'base64+zstd' | 'binary' | 'jsonParsed'];
+              executable: boolean;
+              lamports: number;
+              owner: string;
+              rentEpoch: number;
+              space?: number;
+          } | null)[]
+        | null;
+    err: TransactionError | null;
+    innerInstructions:
+        | {
+              index: number;
+              instructions: (
+                  | {
+                        // Compiled
+                        accounts: number[];
+                        data: string;
+                        programIdIndex: number;
+                        stack_height?: number;
+                    }
+                  | {
+                        // Parsed
+                        parsed: unknown;
+                        program: string;
+                        program_id: string;
+                        stack_height?: number;
+                    }
+                  | {
+                        // PartiallyDecoded
+                        accounts: string[];
+                        data: string;
+                        program_id: string;
+                        stack_height?: number;
+                    }
+              )[];
+          }[]
+        | null;
+    logs: string[] | null;
+    returnData: {
+        data: [string, 'base64'];
+        programId: string;
+    } | null;
+    unitsConsumed: number | null;
+}
+
+export function getSolanaErrorFromJsonRpcError({ code, data, message }: JsonRpcError): SolanaError {
+    if (code === -32002 /* JSON_RPC_SERVER_ERROR_SEND_TRANSACTION_PREFLIGHT_FAILURE */) {
+        const { err, ...preflightErrorContext } = (data ?? {}) as RpcSimulateTransactionResult;
+        return new SolanaError(SOLANA_ERROR__JSON_RPC__SERVER_ERROR_SEND_TRANSACTION_PREFLIGHT_FAILURE, {
+            ...(err ? { cause: getSolanaErrorFromTransactionError(err) } : null),
+            ...(preflightErrorContext as SolanaErrorContext[typeof SOLANA_ERROR__JSON_RPC__SERVER_ERROR_SEND_TRANSACTION_PREFLIGHT_FAILURE]),
+        });
```

Comment:
I'd love it if we could add some additional context data:

```ts
{
    failedInstruction: null | {
        index: number;
        programErrorCode: number;
        programAddress: number;
    }
}
```

Which we would parse ourselves here:

```ts
{
    // ...
   ...(isInstructionError(logs) ? parseInstructionError(logs) : null)
}
```

The logic here would most be the same as the one in `@solana/programs`.

Reply:
I _think_ this already gives you that.

* Preflight fails. You get a preflight failed error
* The `cause` gets set to the result of `getSolanaErrorFromTransactionError()`
* Because the transaction error is _specifically_ an instruction error, it delegates to `getSolanaErrorFromInstructionError()`

So at the end you essentially get:

```ts
const e = new SolanaError(
    SOLANA_ERROR__JSON_RPC__SERVER_ERROR_SEND_TRANSACTION_PREFLIGHT_FAILURE,
    {
        cause: new SolanaError(
            SOLANA_ERROR__INSTRUCTION_ERROR__CUSTOM,
            { code: 123, index: 456 },
        ),
    },
);
```

You can decode the error using something like:

```ts
if (isSolanaError(e.cause, SOLANA_ERROR__INSTRUCTION_ERROR__CUSTOM)) {
    const { code, index } = e.cause.context;
    const { programAddress } = transaction.instructions[index];
    const customError = decodeThisShit(programAddress, code);
}
```

Reply:
Oh I see, that's great then!